lib/AssemblyUtil/TypeToFasta.py

```diff
@@ -0,0 +1,80 @@
+        self.ws_url = ws_url
+        self.callback_url = callback_url
+        self.scratch = scratch
+        self.dfu = DataFileUtil(callback_url)
+        self.mgu = MetagenomeUtils(callback_url)
+
+    def type_to_fasta(self, ctx, ref_lst):
+
+        # Initiate objects
+        fasta_dict, fasta_array = dict(), []
+        atf = AssemblyToFasta(self.callback_url, self.scratch)
+
+        # Get type info for each ref in ref_lst
+        for idx, ref in enumerate(ref_lst):
+            # Initiate objects and get KBase object type with get_object_info3
+            upas = []
+            obj = {"ref": ref}
+            obj_info = self.ws_url.get_object_info3({"objects": [obj]})
+            obj_type = obj_info["infos"][0][2]
+
+            # Find type of object and get back KBase object with dfu.get_objects
+            if 'KBaseSets.GenomeSet' in obj_type:
+                obj_data = self.dfu.get_objects({"object_refs": [ref]})['data'][0]
+                upas = [gsi['ref'] for gsi in obj_data['data']['items']]
+            elif 'KBaseSearch.GenomeSet' in obj_type:
+                obj_data = self.dfu.get_objects({"object_refs": [ref]})['data'][0]
+                upas = [gse['ref'] for gse in obj_data['data']['elements'].values()]
+            elif "KBaseGenomes.Genome" in obj_type:
+                upas = [ref]
+            # If type Assembly, ContigSet or AssemblySet use assembly_as_fasta
+            elif "KBaseGenomes.ContigSet" in obj_type or "KBaseGenomeAnnotations.Assembly" in obj_type:
+                faf = [atf.assembly_as_fasta(ctx, obj)]
+                fasta_array.extend([faf[0]['path'], ref])
+
+            elif "KBaseSets.AssemblySet" in obj_type:
+                fasta_paths = []
+
+                obj_data = self.dfu.get_objects({"object_refs": [ref]})['data'][0]
+
+                for item_upa in obj_data['data']['items']:
+                    faf = [atf.assembly_as_fasta(ctx, {"ref": item_upa['ref']})]
+                    fasta_paths.extend([faf[0]['path'], item_upa['ref']])
+                    fasta_array = fasta_paths
+
+            elif 'KBaseMetagenomes.BinnedContigs' in obj_type:
+                fasta_paths = []
+
+                bin_file_dir = self.mgu.binned_contigs_to_file({'input_ref': ref, 'save_to_shock': 0})['bin_file_directory']
+                for (dirpath, dirnames, filenames) in os.walk(bin_file_dir):
+                    for fasta_file in filenames:
+                        fasta_path = os.path.join(self.scratch, fasta_file)
+                        copyfile(os.path.join(bin_file_dir, fasta_file), fasta_path)
+                        fasta_paths.extend([fasta_path, ref])
+                    break
+                fasta_array = fasta_paths
+            # if statement for genome upa
+            if upas:
+                for genome_upa in upas:
+                    genome_data = self.ws_url.get_objects2({'objects': [{"ref": genome_upa}]})['data'][0]['data']
```

Comment:
Here you might want to consider a way to get the refs without actually pulling the genome fully as they can be large. This pulls gigs of data to get a few k, if that. You may also want to batch this up.

Reply:
Sadly, I need to use get objects2 here because I either need the 'assembly_ref', or the list of 'object_refs' for GenomeSet and/or AssemblySet objects, which are only found in the 'data' dictionary of the retrieved object. 

I would use get_object_info3 if I could but the object info dictionary doesn't contain the assembly_ref or much else really.

Reply:
All the options I outlined earlier, including batching mentioned here, would be implemented with `get_objects2`

Reply:
earlier options: https://github.com/kbaseapps/AssemblyUtil/pull/27#discussion_r293568925

Reply:
The second one is 2 steps, GO2 then GOI3

Reply:
Fixed this by:
genome_data = self.ws.get_objects2({'objects': [{"ref": genome_upa, 'included' : ['/assembly_ref/']}]})['data'][0]['data']

Only doing this for Genome objects though. Not assemblyset or genomeset objects.

Reply:
Do you not need to worry about `config_ref`?

Reply:
contig_ref added.